BootloaderCommonPkg/Library/ShellLib/CmdBoot.c

```diff
@@ -266,26 +266,38 @@ GetBootFileInfo (
   EFI_STATUS                 Status;
   UINTN                      Length;
 
-  do {
-    ShellPrint (L"Enter file path string (max length of %d)\n",
-                sizeof (BootOption->Image[0].FileName) - 1
-                );
-    ShellPrint (L"(default '%a') ", CurrOption->Image[0].FileName);
-    Status = ShellReadLine (Shell, Buffer, BufferSize);
-    if (EFI_ERROR (Status)) {
-      return Status;
-    }
-    Length = StrLen (Buffer);
-    if (Length == 0) {
-      CopyMem (BootOption->Image[0].FileName, CurrOption->Image[0].FileName, sizeof (CurrOption->Image[0].FileName));
-      break;
-    }
-    if (Length < sizeof (BootOption->Image[0].FileName)) {
-      UnicodeStrToAsciiStrS (Buffer, (CHAR8 *)BootOption->Image[0].FileName, sizeof (BootOption->Image[0].FileName));
-      break;
+    for (int image_index = 0; image_index < LoadImageTypeMax; image_index++){
+      if (CurrOption->Image[image_index].FileName[0] != '\0' && image_index != LoadImageTypeMisc){
+        do {
+          if (image_index == 0){
+            ShellPrint (L"Enter file path string (max length of %d)\n",
+                        sizeof (BootOption->Image[0].FileName) - 1
+                        );
+          }
+          else {
+            ShellPrint (L"Enter file path string for extra image %d (max length of %d)\n",
+                        image_index,
+                        sizeof (BootOption->Image[0].FileName) - 1
+                        );
+          }

```

Comment:
from BootloaderCommonPkg/Include/Guid/OsBootOptionGuid.h, Image[1] is not an extra Image.
  // Image[0] is for normal boot OS
  // Image[1] is for Pre OS
  // Image[3-6] is for extra Images

Reply:
This fix is for PerOS as well. I'll modify the commit message for better explaination.

Reply:
Not about commit message. you didn't prompt user to enter PreOS image from the command line. and you didn't check the boot flag before asking for PreOS and Extra Image.

Reply:
Noted and submitted patch

Reply:
I didn't see the flag is checked. Why we always ask user input PreOs if the PreOS flag is not set?
This is the preferred update:
Update GetBootFileInfo () to add ImageIndex parameter so that it could be re-used for all images.
For PreOS image:
if ((CurrOption->BootFlags & BOOT_FLAGS_PREOS) != 0) {
ShellPrint (L"PreOS flag is set, Get PreOS image:\n ");
Status = GetBootFileInfo (Shell, Buffer, sizeof (Buffer), &BootOption, CurrOption, LoadImageTypePreOs);
if (EFI_ERROR (Status)) {
goto ExitBootCmd;
}

Similar we could get extra image name when extra flag is set. It is fine to only get 1 extra image for Image[3] since other extra images are not really used.